extensions/sockets/src/test/java/org/mule/extension/socket/TcpReconnectionTestCase.java

```diff
@@ -0,0 +1,140 @@
+
+    @Mock
+    private ConnectionSettings connectionSettings;
+
+    private TcpServerSocketProperties serverSocketProperties;
+    private TcpClientSocketProperties clientSocketProperties;
+
+    @Mock
+    private MessageHandler handler;
+
+    private TcpListenerConnection listenerConnection;
+
+    private TcpRequesterConnection requesterConnection;
+
+    private String host;
+    private int port;
+
+    @Before
+    public void before() throws ConnectionException
+
+    {
+        port = dynamicPort.getNumber();
+        host = "localhost";
+        when(connectionSettings.getPort()).thenReturn(port);
+        when(connectionSettings.getHost()).thenReturn(host);
+        when(connectionSettings.getInetSocketAddress()).thenReturn(new InetSocketAddress(host, port));
+
+        // Mockito returns 0 in mocked boxed types, that leads to invalid socket configurations
+        serverSocketProperties = new TcpServerSocketProperties();
+        clientSocketProperties = new TcpClientSocketProperties();
+
+        listenerConnection = new TcpListenerConnection(connectionSettings, new SafeProtocol(), serverSocketProperties, new TcpServerSocketFactory());
+        requesterConnection = new TcpRequesterConnection(connectionSettings, new ConnectionSettings(), new SafeProtocol(), clientSocketProperties, new TcpSocketFactory());
+    }
+
+    @Test
+    public void failOnInvalidPort() throws Exception
+    {
+        int invalidPort = -1;
+        connectionSettings = new ConnectionSettings(invalidPort, host);
+        listenerConnection = new TcpListenerConnection(connectionSettings, new SafeProtocol(), serverSocketProperties, new TcpServerSocketFactory());
+        expectedException.expect(IllegalArgumentException.class);
+        expectedException.expectMessage(format("port out of range:%d", invalidPort));
+        listenerConnection.connect();
+    }
+
+    @Test
+    public void failToBindListenerInAnOccupiedPort() throws Exception
+    {
+        expectedException.expect(ConnectionException.class);
+        expectedException.expectMessage(format("Could not bind socket to host '%s' and port '%d'", host, port));
+        listenerConnection.connect();
+
+        TcpListenerConnection secondListener = new TcpListenerConnection(connectionSettings, new SafeProtocol(), serverSocketProperties, new TcpServerSocketFactory());
+        secondListener.connect();
+    }
+
+    @Test
+    public void requesterFailsToConnect() throws Exception
+    {
+        expectedException.expect(ConnectionException.class);
+        expectedException.expectMessage(format("Could not connect TCP requester socket to host '%s' on port '%d'", host, port));
+        requesterConnection.connect();
+    }
+
+    @Test
+    public void closeListener() throws Exception
+    {
+        expectedException.expect(SocketException.class);
+        expectedException.expectMessage("Socket is closed");
+        listenerConnection.connect();
+
+        new Runnable()
```

Comment:
besides if you were doing it in a different thread, you need a latch to control race conditions. This test is not doing what you say it does, and if it did, it'd be flaky